aten/src/ATen/autocast_mode.cpp

```diff
@@ -9,6 +9,13 @@
 #include <iostream>
 #include <exception>
 
+// pulls in AT_CUDNN_ENABLED() as defined by cmake
+#include <ATen/cuda/CUDAConfig.h>
+
+#if AT_CUDNN_ENABLED()
+#include <ATen/native/cudnn/RNNUtils.h>
```

Comment:
Danger danger will robinson. `ATen/autocast_mode.cpp` is compiled as part of ATen_cpu and it should not access any headers in the CUDA directory. You will probably have to chuck these autocast wrappers in a separate file in `ATen/cuda`

Reply:
hmmm but if I move my wrapper as an inline function in a header (or header + cpp) in ATen/cuda, then include the header in autocast_mode.cpp, seems like I'm no better off.

Reply:
Do you mean I should put my wrapper declaration+definition in, say, `ATen/cuda/AutocastRNN.h`+`.cpp`,
have `ATen/autocast_mode.cpp` include `AutocastRNN.h`,
and have `ATen/cuda/AutocastRNN.cpp` be the thing that includes `ATen/native/cudnn/RNNUtils.h`?
That would mean `ATen/native/cudnn/RNNUtils.h` doesn't wind up directly included in `autocast_mode.cpp`.  Or am I misinterpreting?

Reply:
Also, are the files in `ATen/cuda` compiled into a separate library from ATen_cpu?  If so, and I do the above, should I declare `_cudnn_rnn_cast_reflatten` with `TORCH_CUDA_API`?

Would doing the above using `ATen/cudnn` instead of `ATen/cuda` to host my files be equally valid?  If so,  I think `cudnn` rather than `cuda` is a more appropriate home.

Reply:
You should do the registration inside the cpp file in cuda. Then you do not need to include the header from `ATen/autocast_mode.cpp`

Reply:
Feel free to put it in cudnn directory; both dirs end up in torch_cuda library in the end.

Reply:
hopefully resolved